filecoin-proofs/src/api/update.rs

```diff
@@ -0,0 +1,616 @@
+            <Tree::Hasher as Hasher>::Domain::try_from_bytes(&p_aux.comm_c.into_bytes())?,
+            <Tree::Hasher as Hasher>::Domain::try_from_bytes(&p_aux.comm_r_last.into_bytes())?,
+            &new_replica_path,
+            &new_cache_path,
+            &sector_key_path,
+            &sector_key_cache_path,
+            &staged_data_path,
+            u64::from(HSelect::from(porep_config)) as usize,
+        )?;
+
+    comm_d_domain.write_bytes(&mut comm_d)?;
+    comm_r_domain.write_bytes(&mut comm_r)?;
+    comm_r_last_domain.write_bytes(&mut comm_r_last)?;
+
+    ensure!(comm_d != [0; 32], "Invalid all zero commitment (comm_d)");
+    ensure!(comm_r != [0; 32], "Invalid all zero commitment (comm_r)");
+    ensure!(
+        comm_r_last != [0; 32],
+        "Invalid all zero commitment (comm_r)"
+    );
+    ensure!(
+        verify_pieces(&comm_d, piece_infos, porep_config.into())?,
+        "pieces and comm_d do not match"
+    );
+
+    info!("encode_into:finish");
+    Ok((comm_r, comm_r_last, comm_d))
+}
+
+/// Reverses the encoding process and outputs the data into out_data_path.
+#[allow(clippy::too_many_arguments)]
+pub fn decode_from<Tree: 'static + MerkleTreeTrait<Hasher = TreeRHasher>>(
+    porep_config: PoRepConfig,
+    out_data_path: &Path,
+    replica_path: &Path,
+    sector_key_path: &Path,
+    sector_key_cache_path: &Path,
+    comm_d_new: Commitment,
+) -> Result<()> {
+    info!("decode_from:start");
+
+    // NOTE: p_aux has comm_c and comm_r_last
+    let p_aux: PersistentAux<<Tree::Hasher as Hasher>::Domain> = {
+        let p_aux_path = sector_key_cache_path.join(CacheKey::PAux.to_string());
+        let p_aux_bytes = fs::read(&p_aux_path)
+            .with_context(|| format!("could not read file p_aux={:?}", p_aux_path))?;
+
+        deserialize(&p_aux_bytes)
+    }?;
+
+    let nodes_count = u64::from(porep_config.sector_size) as usize / NODE_SIZE;
+    EmptySectorUpdate::<Tree>::decode_from(
+        nodes_count,
+        out_data_path,
+        replica_path,
+        sector_key_path,
+        sector_key_cache_path,
+        <Tree::Hasher as Hasher>::Domain::try_from_bytes(&p_aux.comm_c.into_bytes())?,
+        comm_d_new.into(),
+        <Tree::Hasher as Hasher>::Domain::try_from_bytes(&p_aux.comm_r_last.into_bytes())?,
+        u64::from(HSelect::from(porep_config)) as usize,
+    )?;
+
+    info!("decode_from:finish");
+    Ok(())
+}
+
+/// Removes encoded data and outputs the sector key.
+#[allow(clippy::too_many_arguments)]
+pub fn remove_encoded_data<Tree: 'static + MerkleTreeTrait<Hasher = TreeRHasher>>(
+    porep_config: PoRepConfig,
+    sector_key_path: &Path,
+    sector_key_cache_path: &Path,
+    replica_path: &Path,
+    replica_cache_path: &Path,
+    data_path: &Path,
+    comm_d_new: Commitment,
+) -> Result<()> {
+    info!("remove_data:start");
+
+    // NOTE: p_aux has comm_c and comm_r_last
+    let p_aux: PersistentAux<<Tree::Hasher as Hasher>::Domain> = {
```

Comment:
This recovery of the different files is done quite often, maybe this could be extracted into a reusable function, sharing some code?